### ISSUE TITLE
JSpecify: iterating over a optionally nullable iterable

### ISSUE THREAD

**Issue by unv-unv** [uber/NullAway#1799]
```java
@NullMarked
public class Test {
    public final <E extends @Nullable Object, K, V extends @Nullable Object>
    Map<K, V> toMap(Iterable<E> elems, Function<E, K> toKey, Function<E, V> toValue) {
        Map<K, V> map = new LinkedHashMap<>();
        for (E e : elems) {
            // BUG: apply(e): [NullAway] passing @Nullable parameter 'e' where @NonNull is required
            map.put(toKey.apply(e), toValue.apply(e));
        }
        return map;
    }
}
```

The following compilation error is shown:
```
[ERROR] Test.java:[8,32] [NullAway] passing @Nullable parameter 'e' where @NonNull is required
[ERROR]     (see http://t.uber.com/nullaway )
[ERROR] Test.java:[8,50] [NullAway] passing @Nullable parameter 'e' where @NonNull is required
[ERROR]     (see http://t.uber.com/nullaway )
```

JSpecify mode is on.

NullAway 0.14.1, Error Prone 2.50.0

**Comment by Artur-** [comment]
This seems like a regression in 0.14.1 that originates in #1748 and #1749

**Comment by msridhar** [comment]
Thanks for the report.  This is #1727.  I was hoping the impact of that issue was limited to JSpecifyExperimental, but it appears it is not.  #1748 and #1749 are relevant since we're doing more reasoning about the types of vars declared in foreach loops.

Fixing #1727 is a high priority.  For now you'll have to suppress these warnings, sorry.  I'll make sure this scenario passes when we fix #1727.
